tap_recharge/client.py

```diff
@@ -44,59 +42,129 @@ class RechargeMethodNotAllowedError(RechargeError):
+    500: {
+        "exception": RechargeInternalServiceError,
+        "message": "The request could not be processed due to internal server error.",
+    },
+    501: {
+        "exception": RechargeUnimplementedResourceError,
+        "message": "The resource requested has not been implemented in the current version.",
+    },
+    503: {
+        "exception": RechargeThirdPartyServiceTimeoutError,
+        "message": "A third party service on which the request depends has timed out.",
+    }
+}
 
 def get_exception_for_error_code(error_code):
-    return ERROR_CODE_EXCEPTION_MAPPING.get(error_code, RechargeError)
+    """Function to retrieve exceptions based on error code"""
+    if error_code == 429:
+        # Delay for 5 seconds for leaky bucket rate limit algorithm
+        time.sleep(5)
+
+    exception = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get('exception')
+    if not exception:
+        if error_code >= 500:
+            return Server5xxError
+        return RechargeError
+    return exception
 
 def raise_for_error(response):
+    """Function to raise custom error along with message based on response."""
+
+    error_code = response.status_code
+
     try:
-        response.raise_for_status()
-    except (requests.HTTPError, requests.ConnectionError) as error:
-        try:
-            content_length = len(response.content)
-            if content_length == 0:
-                # There is nothing we can do here since Recharge has neither sent
-                # us a 2xx response nor a response content.
-                return
-            response = response.json()
-            if ('error' in response) or ('errorCode' in response):
-                message = f"{response.get('error', str(error))}: \
-                    {response.get('message', 'Unknown Error')}"
-                error_code = response.get('status')
-                ex = get_exception_for_error_code(error_code)
-                if response.status_code == 401 and 'Expired access token' in message:
-                    LOGGER.error("Your access_token has expired as per Recharge’s security \
-                        policy. \n Please re-authenticate your connection to generate a new token \
-                        and resume extraction.")
-                raise ex(message)
-            else:
-                raise RechargeError(error)
-        except (ValueError, TypeError):
-            raise RechargeError(error)
+        response_json = response.json()
+    except ValueError:
+        response_json = {}
+
+    error_message = response_json.get('error', response_json.get('errors'))
+
+    if not error_message:
+        error_message = ERROR_CODE_EXCEPTION_MAPPING.get(error_code, {}).get('message', 'Unknown Error')
+
+    if error_code == 401 and 'bad authentication' in error_message:
+        LOGGER.error("Your API Token has been deleted or the token is invalid.\n Please re-authenticate your connection to generate a new token and resume extraction.")
```

Comment:
This must be just before the exception so move it 2 lines down.

Reply:
Updated the location of the logger.